feat: GraphQL Code Generator preset-client SWC plugin post (#1110)
the-guild-org/website: website/pages/blog/optimize-bundle-size-with-swc-plugin-for-graphql-codegen.mdx

---
title: Optimize Bundle Size with SWC Plugin for GraphQL Code Generator
authors: jessevelden
tags: [client-preset, codegen]
date: 2023-02-24
description:
  'Optimize your Bundle Size with the SWC plugin for client-preset for GraphQL Code Generator'
# image: /blog-assets/optimize-bundle-size-with-swc-plugin-for-graphql-codegen/coverTODO.png
# thumbnail: /blog-assets/optimize-bundle-size-with-swc-plugin-for-graphql-codegen/thumbnailTODO.png
---

With the new
[`client-preset` for GraphQL Code Generator](https://graphql-code-generator.com/plugins/presets/client-preset)
it is possible to write your GraphQL queries inlined, next to your component in a type-safe way.

> This can greatly improve the Developer Experience (DX).

Reviewer: Maybe lets link to https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen here?

and use the `graphql` function to parse the operation and get the correct types.

```tsx
// Instead of operations in a separate .graphql file and importing generated hooks:
import { useGetCartByIdQuery } from './graphql/generated-hooks'
const { data } = useGetCartByIdQuery({ variables: { id: '1' } })

// You can write your operations directly next to your component and get the correct types:
import { graphql } from './graphql'
const GetCartByIdDocument = graphql(/** GraphQL */ ` query GetCartById($id: ID!) {
   id
   items { 
      id
      ...CartItem
   }
 }`)
 // Awesome! We have the correct types for this operation even when we specify it next to our component!
```

However, this approach can come at a cost - the bundle size of your pages can increase
significantly, especially as the number of queries and mutation grows. In this post, we will show
you what the problem is and how to optimize your bundle size with the new
[`@graphql-codegen/client-preset-swc-plugin`](https://www.npmjs.com/package/@graphql-codegen/client-preset-swc-plugin)
plugin for your applications supported by the SWC compiler such as projects using

> [Next.js](https://nextjs.org).

Reviewer: ```suggestion
[Next.js](https://nextjs.org) or [Vite](https://vitejs.dev/).
```

The generated code output for the `graphql` function is a function that looks similar to
the following code, where code-splitting and tree-shaking is not easily possible.

```ts
const documents = {
  "fragment CartItems on CartItem {\n name\n quantity \n  }\n  ": types.CartItemsFragmentDoc,
   "query GetCartById($id: ID!) {\n cart(id: $id) {\n id \n items { \n id \n ...CartItem\n }\n totalItems\n}\n  }\n": types.GetCartByIdDocument,
};
export function graphql(source: "fragment CartItems on CartItem {\n name\n quantity\n }\n "): (typeof documents)["\n fragment CartItem on CartItem {\n name\n quantity\n }\n "];
export function graphql(source: "query GetCartById($id: ID!) {\n cart(id: $id) {\nid\nitems {\nid\n...CartItem\n}\ntotalItems\n }\n }\n"): (typeof documents)["\n query GetCartById($id: ID!) {\n cart(id: $id) {\n id\n items {\n id \n ...CartItem \n } \n totalItems \n } \n }\ n"];
export function graphql(source: string): unknown;
export function graphql(source: string) {
  // Here we access the documents object, so the bundler cannot make assumption over what is
  //  actually used at runtime
  return (documents as any)[source] ?? {};
}

> /**

Reviewer: This comment seems to be an accident.
```suggestion

```

```

This means that every time you use the `graphql` function, you are importing basically all used
operations in your project, even if those are not being used. This can be a problem if you have a

> lot of queries and mutations in your project.

Reviewer: ```suggestion
lot of documents (query, mutation and fragments) in your project.
```

For a more information/ history of this functionality, you can read more about it in the former 
[`gql-tag-operations-preset`](https://the-guild.dev/graphql/codegen/plugins/presets/gql-tag-operations-preset) that 
is now replaced by the `client-preset`. This idea was originally explored by [Maël Nison](https://github.com/arcanis) in 

> [this Github project](https://github.com/arcanis/graphql-typescript-integration).

Reviewer: We can wrap this in an `Callout` component

## The Solution: A Compiler Plugin

> In order to have type safety in This is where a compiler can make a big difference. Instead of using

Reviewer: ```suggestion
This is where a compiler can make a big difference. Instead of using
```

Seems like a typo
